src/Simulation/TargetDefinitions/Decompositions/MeasureWithNoReuse.qs

```diff
@@ -39,27 +40,120 @@ namespace Microsoft.Quantum.Intrinsic {
     /// If the basis array and qubit array are different lengths, then the
     /// operation will fail.
     operation Measure (bases : Pauli[], qubits : Qubit[]) : Result {
-        if (Length(bases) != Length(qubits)) { fail "Arrays 'bases' and 'qubits' must be of the same length."; }
-        if (Length(bases) == 1) {
-            // Because the qubit cannot be reused after measurement, there is no
-            // need to unprepare the Pauli mapping.
-            MapPauli(qubits[0], PauliZ, bases[0]);
-            return M(qubits[0]);
+        if Length(bases) != Length(qubits) { fail "Arrays 'bases' and 'qubits' must be of the same length."; }
+        if Length(bases) == 0 {
+            return Zero;
         }
-        else {
-            use q = Qubit() {
-                within {
-                    H(q);
-                }
-                apply {
-                    for k in 0 .. Length(bases) - 1 {
-                        if( bases[k] == PauliX ) { Controlled X ([q], qubits[k]); }
-                        if( bases[k] == PauliZ ) { Controlled Z ([q], qubits[k]); }
-                        if( bases[k] == PauliY ) { Controlled Y ([q], qubits[k]); }
-                    }
-                }
-                return M(q);
+        elif Length(bases) == 1 {
```

Comment:
The only difference between this version of measure and the other one seems to be when the length of bases is 1. When one decomposition would be used over the other?

Reply:
When qubits can be reused after measurement, `Measure` can operate directly on the target qubit and allow later operations. When the platform does not allow reuse of a qubit after measurement, the alternate decomposition uses an auxiliary and entanglement to get information about the current state (and collapse the target) while still allowing it to be used in later operations. I've added a comment explaining this succinctly, but I'm debating whether or not this needs a longer explanation somewhere...

Reply:
I think the explanation you added could be good enough for now, and the name of the source code file also provides a good hint on what that measure decomposition is about :)